### DISCUSSION TITLE
Is it possible to use a Layout?

### DISCUSSION THREAD

**Discussion by DumboJet** [dustinchilson/NLog.Targets.GraylogHttp#145] [Q&A]
The `GraylogHttpTarget` class contains a `Layout` property.
Any chance I can use it to render some extra properties automatically, instead of manually doing this? :

```
var logger = LogManager.GetCurrentClassLogger();
var e = new LogEventInfo(LogLevel.Fatal, "Test", "Test Message");
e.Properties["test_prop"] = "test property";
logger.Log(e);
```

I tried to use the Layout property, but I don't think it works.

Essentially, what I want is this code:

```
logger.Log("Error");
```

...to log some extra properties using a layout (without modifications to the code above).
Not sure if there is a good way to transform a layout into the properties of `LogEventInfo` and supply them behind the scenes.

**Comment by dustinchilson** [comment]
You should be able to, but you shouldn't need to use the layout. you do need to set `IncludeEventProperties = true`. I have not tested this extensively.

See PR #10

**Comment by DumboJet** [comment]
But `IncludeEventProperties` requires me to manually set the properties on `LogEventInfo`, correct?
I want to avoid this.
My solution uses a layout to render some parameters to a file target and I will need something similar that works globally with GrayLog as well, without me changing each and every logging location in the code.

I have tried `IncludeEventProperties = true` and it didn't add anything to the logs. I guess it expects me to set the properties manually on each logging location in the code, as shown in my first code snippet. But I want to avoid this.

I guess, I could inherit `GraylogHttpTarget` and override the `void Write(LogEventInfo logEvent)` method to automatically include the properties on all logging operations...

If you have any better ideas, let me know.

**Comment by DumboJet** [comment]
Here is what I did and it works:

```

    public class MyGraylogHttpTarget : GraylogHttpTarget
    {
        private readonly ILogDataRepositoryFactory _logDataRepoFactory;

        public MyGraylogHttpTarget(ILogDataRepositoryFactory logDataRepoFactory)
        {
            _logDataRepoFactory = logDataRepoFactory;
        }

        protected override void Write(LogEventInfo logEvent)
        {
            if (logEvent == null)
            {
                return;
            }

            // Add your custom properties.
            var repo = _logDataRepoFactory.GetLogDataRepository();
            var logData = repo.GetActiveRequestLogData();
            logEvent.Properties.Add("MyProperty", logData.MyProperty);

            base.Write(logEvent);
        }
    }
```

Then, I use this target, instead of the default one.

**Comment by dustinchilson** [comment]
You'd specify the properties you want sent to graylog in the config 

like this
```
<parameter name="orderid" layout="${event-context:item=orderid}" />
<parameter name="user" layout="${event-context:item=user}" />
```

then use the structured logging to provide the data. 
```
logger.Info("Order {orderid} created for {user}", 42, "Kenny");
```
https://github.com/NLog/NLog/wiki/How-to-use-structured-logging

Which is effectively the same thing as setting the property, but also interpolates it into the message.

**Comment by dustinchilson** [comment]
Your solution for global parameters is great, Missed that part.

But yes each log message does need the params attached somehow. Serilog has enrichers which work the way you want but I'm not sure NLog has an equivalent.

**Comment by dustinchilson** [comment]
another alternative if the values are static

```
<parameter name="environment" layout="production" />
```

**Comment by DumboJet** [comment]
Thanks for the responses. :)
My parameters are dynamic (RequestId, UserId, etc), so I think I will stick with the class I have created above for now.
But, if you want, maybe you can provide this functionality on your `GraylogHttpTarget`.
E.g., have another `GraylogHttpTarget` constructor that accepts a callback for getting a dictionary of property key/value pairs.
